### DISCUSSION TITLE
BUG: Scoped component not updating properties on parent changes

### DISCUSSION THREAD

**Discussion by verybigelephants** [vuejs/petite-vue#228] [Q&A]
Hello,

could please someone explain what am i doing wrong?

I have a parent component, let's say "Library", iterating over array of children, let's say "Books". However, when i update the children Books array petite vue doesn't catch its changes, only if i delete the array, wait a bit/on next tick, and then re-assign the books again.

But then again, it catches the change when i just `.concat()` the array

This is the full (non) working example to explain what i mean:

```
<!doctype html>
<html lang="en">
<head>
	<meta charset="utf-8">
	<script src="https://unpkg.com/petite-vue" defer init></script>
</head>
<body>
	<div v-scope="Library({books:books_1})" @vue:mounted="mounted">
		<ul>
			<li 
				v-for="(book, i) in books"
				v-scope="Book({title:book.title})"
			>
				<span v-text="title"></span>
			</li>
		</ul>
		
		<button @click="updateBooks">Update books</button>
		<button @click="updateBooksNextTick">Update books on next tick</button>
		<br /> <br />
		<button @click="thisIsStrange">But it catches the books update when i concat it</button>
	</div>
	
	<script type="text/javascript">
		var books_1 = [
			{title:"test 1"},
			{title:"test 2"},
			{title:"test 3"}
		];
	
		function Library(props){
			return {
				books:props.books,
		
				updateBooks(){
					this.books = this.randBooks();
				},
				updateBooksNextTick(){
					let self = this;
					this.books = [];  
					//this lets petite vue to realize that all the children are gone or something
					this.$nextTick(() => {
						this.books = self.randBooks();
					});
				},
				thisIsStrange(){
					this.books = this.books.concat(this.randBooks() );
				},
				
				randBooks(){
					let collection = [];
					for(var i=0;i<3;i++){
						collection.push({title:(Math.random().toString(36).substr(2, 5))});
					}
					return collection;
				},
				mounted: function(){
					console.log("mounted", this.books);
				}
			}
		}
		function Book(props){
			return {
				title: props.title
			}
		}
		
		document.addEventListener("DOMContentLoaded", function(event) {
			PetiteVue.createApp({
				
			}).mount('body');
		});

	</script>
</body>
</html>
```

What is going on here? What am i doing wrong?

**Comment by sqllyw** [comment]
try reactive:
    
        var books_1 = reactive( [
            {title:"test 1"},
            {title:"test 2"},
            {title:"test 3"}
        ])

**Comment by verybigelephants** [comment]
@sqllyw that doesn't make any sense, that would only make specific on-the-fly array of items book items reactive, not the actual books_1. And even then, this is just a property that gets passed down to Library. 

(anyways i did try to use  `global = PetiteVue.reactive({ books: books_1});` on a the input and then `Library({books:global.books})` in the scope builder and it didn't help this issue of course)

**Comment by sqllyw** [comment]
ok, the issue seems related to the Book component, when you change the book in :

    this.books = this.randBooks();

v-scope="Book({title:book.title})" is still referring to the value that you passed from books_1, because props are not responsive, this is very similar to https://github.com/vuejs/petite-vue/discussions/109, the selected answer in 109 explains why, and also offer a solution, to apply to your example:

~~~~
<ul>
    <li v-for="(book, i) in books" v-scope="Book({title: () => book.title})">
        <span v-text="title()"></span>
    </li>
</ul>
~~~~

as for why this works:
~~~~
    this.books = [];
     //this lets petite vue to realize that all the children are gone or something
     this.$nextTick(() => {
     this.books = self.randBooks();
     });
~~~~

your comment already explains it.

**Reply by verybigelephants** [reply]
i will be honest, i am not 100% sure why passing an anonymous functoin like `title: () => book.title` solves the issue, and using this seems like a bad practice/unexpected component structure in case someone else wanted to re-use the Book component in the future

in the end i ended up removing the "Book" component altogether and i just iterate items from the global books object `v-for="(book, i) in books"` and then `<span v-text="book.title"></span>` directly

while this doesn't seem very clean, not having a book component at all, but this seems like a better tradeoff than trying to hack this unexpected behavior in terms of code manageability, becuases this really feels like a petite vue bug

**Comment by sqllyw** [comment]
in your example, the Book component got passed a primitive value, if you change  the books_1[0].title it will not be reflected since primitive data is non reactive by design, you can pass a object, then change the object properties, it is reactive, here is a slight change to your code that illustrate this point, you can still have a good code manageability by passing a book object to the Book component:

````
<div v-scope="Library({books:books_1})" @vue:mounted="mounted">
		<ul>
			<li 
				v-for="(book, i) in books"
				v-scope="Book({mybook:book})"
			>
				<span v-text="mybook.title"></span>
			</li>
		</ul>
````
a working sample here:
https://jsfiddle.net/43jkcm8w/

**Reply by verybigelephants** [reply]
Thank you for the suggestion;

however, the issue is the same, as your example doesn't work because you pass the book as an object, but because you are utilizing the `nextTick` in the `updateBooks` method (which would work even if you did `Book({title:book.title})` )

**Reply by sqllyw** [reply]
nextTick is the documented way to update the dom after some state change:
https://vuejs.org/api/general.html#nexttick

**Reply by verybigelephants** [reply]
I know what nextTick is for, i have written regular vue before. And it shall not be called when proprties normally change every time, what would be a a point of a reactive framework like vue then? 

Look, i appreciate you trying to help but form the comments it seems that you fully don't understand what's going on and provided irrelevant code that does not touch this issue, so with all due respect, this is not helpful to anyone.

Also, the application behaves as expected when some component is reading dynamic properties/its changes from a global, not sub-scroped child. Plus, from the first example we may see that when the object in the property updates via  `.concat` petite vue does reflect the change, so this clearly seems like a petite vue bug

**Comment by 604587986** [comment]
You can add `:key="book.title"` on for item

**Reply by verybigelephants** [reply]
that does not do anything regarding the bug

simply - PetiteVue, doesn't reflect reactive variable changes when they are passed down as parameters

**Reply by terryaney** [reply]
@verybigelephants I see you discovered that Array.concat seemed to maintain 'reactivity'.  I'm having a similar problem and could post a new question/code, but have you found a solid way to work with arrays?  I have something like following initialized as 'state' when creating the VUE app:

```
results: {
}
```

Then I need to support all the following scenarios:

1. Assign new property to an array of items.
2. Replace existing property array with new array of items.
3. Merge new array of items into existing property array.
    1. Replacing entire with same 'id', or
    2. Mapping all properties from new element into existing element with same 'id' (adding property or updating value), or
    3. Simply add element if no existing element with same 'id' exists

If the array property exists and I'm adding an element, I switched from `.push` to `.concat` and that kept reactivity working for that scenario.  But I don't know how to solve for the other situations I need.

Been battling this 'bug' for days and could put something more coherent together, but bottom line is, "have you figured out a stable way to work with arrays in reactive state and have 'everything' work in petite-vue (i.e. nested v-scope)?"

**Reply by verybigelephants** [reply]
@terryaney well you can always do any operation on your array and then call `array.concat([])` afterwards. or you may try doing that callback hell nextTick hack after emptying and reassigning the array afterwards

**Comment by verybigelephants** [comment]
there is definitely is some major reactivity bug with this otherwise wonderful library. another issue that might be related to some buggy dom caching:

```
    <div v-if="item_picked">
       <h1 v-html="item_picked.name" @click="unpick()"></h1>
   </div>

    <ul v-for="item in items">
        <li @click="pick(item)">pick me</li>
    </ul>
```
and the pick/unpick functions
```
//some default props for a component
items: props.items,
item_picked:null,
...
pick(item){
    this.item_picked = item;
},
unpick(){
    this.item_picked = null;
}
```

after picking the element and then unpicking it the console throws error `TypeError: item_picked is null` when it tries to access item_picked.name, even though it should not pass the v-if branch

**Reply by terryaney** [reply]
Yes, I've seen that one with v-if's lagging behind a bit and have just gotten used to protecting stuff with `item_picked?.name` in places where that situation arises. :(  Think I'm going to try and make a small reproducible error of my bug and look for any possible workarounds because I can't seem to find any.

**Reply by terryaney** [reply]
Here is the [discussion](https://github.com/vuejs/petite-vue/discussions/169) about the `v-if`.
